src/languageclient.rs

```diff
@@ -395,47 +395,105 @@ impl State {
-                "nvim_buf_add_highlight",
-                json!([
-                    0,
-                    source,
-                    hl_group,
-                    dn.range.start.line,
-                    dn.range.start.character,
-                    dn.range.end.character,
-                ]),
+                "nvim_buf_clear_highlight",
+                json!([0, source, 1, -1]),
             )?;
+            for dn in diagnostics {
+                let severity = dn.severity.unwrap_or(DiagnosticSeverity::Information);
+                let hl_group = diagnosticsDisplay
+                    .get(&severity.to_int()?)
+                    .ok_or_else(|| err_msg("Failed to get display"))?
+                    .texthl
+                    .clone();
+
+                self.call::<_, u8>(
+                    None,
+                    "nvim_buf_add_highlight",
+                    json!([
+                        0,
+                        source,
+                        hl_group,
+                        dn.range.start.line,
+                        dn.range.start.character,
+                        dn.range.end.character,
+                    ]),
+                )?;
+            }
+        } else {
+            self.call::<_, u8>(None, "clearmatches", json!([]))?;
+
+            // Group diagnostics by severity so we can highlight them
+            // in a single call.
+            let mut match_groups: HashMap<_, Vec<_>> = HashMap::new();
+
+            for dn in diagnostics {
+                let severity = dn.severity
+                    .unwrap_or(DiagnosticSeverity::Information)
+                    .to_int()?;
+                match_groups
+                    .entry(severity)
+                    .or_insert_with(Vec::new)
+                    .push(dn);
+            }
+
+            for (severity, dns) in match_groups {
+                let hl_group = diagnosticsDisplay
+                    .get(&severity)
+                    .ok_or_else(|| err_msg("Failed to get display"))?
+                    .texthl
+                    .clone();
+                let ranges: Vec<_> = dns.iter()
+                    .flat_map(|dn| {
+                        if dn.range.start.line == dn.range.end.line {
+                            let length = dn.range.end.character - dn.range.start.character;
+                            // Vim line numbers are 1 off
+                            // `matchaddpos` expects an array of [line, col, length]
+                            // for each match.
+                            vec![[
+                                dn.range.start.line + 1,
+                                dn.range.start.character + 1,
+                                length,
+                            ]]
+                        } else {
+                            let mut middleLines: Vec<_> = (dn.range.start.line + 1
+                                ..dn.range.end.line)
+                                .map(|l| [l + 1, 0, 0])
```

Comment:
According to the doc, when trying to highlight the whole line, 2rd and 3rd item can be omitted. Plus, I'm not 100% sure if `0, 0` will be interpreted as whole line by vim.